### DISCUSSION TITLE
Solver for heat equation with a source term and different boundary conditions

### DISCUSSION THREAD

**Discussion by abhishek-appana** [kinnala/scikit-fem#1088] [Q&A] [answered]
Hello everyone,

I am looking to solve the heat equation on a rectangular geometry with different boundary conditions on the edges as explained below:

<img width="248" alt="image" src="https://github.com/kinnala/scikit-fem/assets/142918023/5547e228-d44b-415d-8b88-1b6a125294e3">

The heat equation is:
$\rho C_p \partial T/\partial t = k \Delta T + Q_{gen} $

The edges 1, 2, 3, 5, 7, 8 have the following boundary condition:
$-k dT/dn = h(T-T_{inf})$
Edges 4 & 6 have the boundary condition:
$-k dT/dn = q_{tabs}$

Below is the code I have written for this. The figure gets generated for t=0, but the solution doesn't get published/generated from the next time step. Can someone please let me know what needs to be changed?

Thank you

```python
from typing import Iterator, Tuple

import numpy as np
from scipy.sparse.linalg import splu

from skfem import *
from skfem.helpers import dot, grad

length = 0.5
height = 0.5
ncells = 50

T_inf = 298
Q_gen = 1000
q_tabs = 500
rho = 25
C_p = 90
k = 25
h = 10

dt = 0.01

mesh = (MeshQuad
        .init_tensor(np.linspace(0, length, ncells),
                     np.linspace(0, height, ncells))
        .with_boundaries({'1': lambda x: np.isclose(x[1],0),
                          '2': lambda x: np.isclose(x[0],0),
                          '3': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0) & np.less_equal(x[0],0.1),
                          '4': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.1) & np.less_equal(x[0],0.2),
                          '5': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.2) & np.less_equal(x[0],0.3),
                          '6': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.3) & np.less_equal(x[0],0.4),
                          '7': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.4) & np.less_equal(x[0],0.5),
                          '8': lambda x: np.isclose(x[0],0.5)}))

e = ElementQuad2()  # or ElementQuad1
basis = {
    'cell': Basis(mesh, e),
    'pos_tab': FacetBasis(mesh, e, facets=mesh.boundaries['4']),
    'neg_tab': FacetBasis(mesh, e, facets=mesh.boundaries['6']),
    'bottom': FacetBasis(mesh, e, facets=mesh.boundaries['1']),
    'left': FacetBasis(mesh, e, facets=mesh.boundaries['2']),
    'top_1': FacetBasis(mesh, e, facets=mesh.boundaries['3']),
    'top_2': FacetBasis(mesh, e, facets=mesh.boundaries['5']),
    'top_3': FacetBasis(mesh, e, facets=mesh.boundaries['7']),
    'right': FacetBasis(mesh, e, facets=mesh.boundaries['8'])}

@BilinearForm
def laplace(u, v, _):
    return dot(grad(u),grad(v))

@BilinearForm
def mass(u, v, _):
    return u*v

@LinearForm
def tabs(v, _):
    return dt*q_tabs*v

@BilinearForm
def rest(u, v, _):
    return dt*h*(u - T_inf)*v

@LinearForm
def loss(v, _):
    return dt*Q_gen*v

T_init = np.repeat(298, basis['cell'].doflocs.shape[1])

M = rho * C_p * asm(mass,basis['cell'])
L = k * asm(laplace, basis['cell'])
rhs_1 = -1 * (asm(tabs,basis['neg_tab']) + asm(tabs, basis['pos_tab']))
rhs_2 = (-1 * (asm(rest,basis['bottom']) + asm(rest,basis['left']) + asm(rest,basis['top_1']) 
              + asm(rest,basis['top_2']) + asm(rest,basis['top_3']) + asm(rest,basis['right'])))
rhs_3 = asm(loss, basis['cell'])

A = M + 0.5 * L * dt
B = M - 0.5 * L * dt

backsolve = splu(A.T).solve

def evolve(t: float, 
           u: np.ndarray) -> Iterator[Tuple[float, np.ndarray]]:
    
    while t < 10:
        t, u = t + dt, backsolve((B @ u) + rhs_1 + rhs_2 + rhs_3)
        yield t, u


if __name__ == '__main__':

    from argparse import ArgumentParser
    from pathlib import Path

    from matplotlib.animation import FuncAnimation
    import matplotlib.pyplot as plt

    from skfem.visuals.matplotlib import plot

    parser = ArgumentParser(description='heat equation in a rectangle')
    parser.add_argument('-g', '--gif', action='store_true', 
                        help='write animated GIF', )
    args = parser.parse_args()

    ax = plot(mesh, T_init[basis['cell'].nodal_dofs.flatten()], shading='gouraud')
    title = ax.set_title('t = 0.00')
    field = ax.get_children()[0]  # vertex-based temperature-colour
    fig = ax.get_figure()
    fig.colorbar(field)


    def update(event):
        t, u = event

        title.set_text(f'$t$ = {t:.2f}')
        field.set_array(u[basis['cell'].nodal_dofs.flatten()])

    animation = FuncAnimation(
        fig,
        update,
        evolve(0., T_init),
        repeat=False,
        interval=50,
    )
    if args.gif:
        animation.save(Path(__file__).with_suffix('.gif'), 'imagemagick')
    else:
        plt.show()
```

**Comment by kinnala** [comment, marked as answer]
This part does not make sense:
```python
@BilinearForm
def rest(u, v, _):
    return dt*h*(u - T_inf)*v
```
It is clearly not bilinear, i.e., linear both in $u$ and $v$.
I suggest you carefully write down the weak formulation of your problem.

**Reply by abhishek-appana** [reply]
Oh jeez! My bad!

I corrected it. The simulation is running now

Thank you so much @kinnala

**Reply by abhishek-appana** [reply]
Hello everyone,

For some reason, the simulation results are not what I expected them to be! After simulating it for 100 seconds, the results I expected should be something like below, with clearly different profile at edges 4 & 6:
![image](https://github.com/kinnala/scikit-fem/assets/142918023/12d00a3e-175b-4f5b-b35d-46291fe71a48)

But this is what I get:
<img width="355" alt="image" src="https://github.com/kinnala/scikit-fem/assets/142918023/db20faf9-70ae-41f0-83a5-970051363d46">

Here is the code I have used. I couldn't figure out if I was implementing something erroneously. Or if the results are like this because of the approximation. Can someone help me figure out what is wrong? Please note that $Q_{gen}$ is uniform throughout the mesh!

```python
from typing import Iterator, Tuple

import numpy as np
from scipy.sparse.linalg import splu

from skfem import *
from skfem.helpers import dot, grad

length = 0.5
height = 0.5
ncells = 50

T_inf = 298
Q_gen = 1000
q_tabs = 500
rho = 25
C_p = 90
k = 25
h = 10

dt = 0.1

mesh = (MeshQuad
        .init_tensor(np.linspace(0, length, ncells),
                     np.linspace(0, height, ncells))
        .with_boundaries({'1': lambda x: np.isclose(x[1],0),
                          '2': lambda x: np.isclose(x[0],0),
                          '3': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0) & np.less_equal(x[0],0.1),
                          '4': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.1) & np.less_equal(x[0],0.2),
                          '5': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.2) & np.less_equal(x[0],0.3),
                          '6': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.3) & np.less_equal(x[0],0.4),
                          '7': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.4) & np.less_equal(x[0],0.5),
                          '8': lambda x: np.isclose(x[0],0.5)}))

e = ElementQuad2()  # or ElementQuad1
basis = {
    'cell': Basis(mesh, e),
    'pos_tab': FacetBasis(mesh, e, facets=mesh.boundaries['4']),
    'neg_tab': FacetBasis(mesh, e, facets=mesh.boundaries['6']),
    'bottom': FacetBasis(mesh, e, facets=mesh.boundaries['1']),
    'left': FacetBasis(mesh, e, facets=mesh.boundaries['2']),
    'top_1': FacetBasis(mesh, e, facets=mesh.boundaries['3']),
    'top_2': FacetBasis(mesh, e, facets=mesh.boundaries['5']),
    'top_3': FacetBasis(mesh, e, facets=mesh.boundaries['7']),
    'right': FacetBasis(mesh, e, facets=mesh.boundaries['8'])}

@BilinearForm
def laplace(u, v, _):
    return dot(grad(u),grad(v))

@BilinearForm
def mass(u, v, _):
    return u*v

@LinearForm
def tabs(v, _):
    return dt*q_tabs*v

@LinearForm
def rest(v, _):
    return dt*h*T_inf*v

@LinearForm
def loss(v, _):
    return dt*Q_gen*v

T_init = np.repeat(298, basis['cell'].doflocs.shape[1])

M = rho * C_p * asm(mass,basis['cell'])
L = k * asm(laplace, basis['cell'])
O = h * (asm(mass,basis['bottom']) + asm(mass, basis['left']) + asm(mass, basis['top_1'])
          + asm(mass, basis['top_2']) + asm(mass, basis['top_3']) + asm(mass, basis['right']))

rhs_1 = -1 * (asm(tabs,basis['neg_tab']) + asm(tabs, basis['pos_tab']))
rhs_2 = (asm(rest,basis['bottom']) + asm(rest,basis['left']) + asm(rest,basis['top_1']) 
        + asm(rest,basis['top_2']) + asm(rest,basis['top_3']) + asm(rest,basis['right']))
rhs_3 = asm(loss, basis['cell'])

A = M + 0.5 * (L + O) * dt
B = M - 0.5 * (L + O) * dt

backsolve = splu(A.T).solve

def evolve(t: float, 
           u: np.ndarray) -> Iterator[Tuple[float, np.ndarray]]:
    
    while t < 100:
        t, u = t + dt, backsolve((B @ u) + rhs_1 + rhs_2 + rhs_3)
        yield t, u

if __name__ == '__main__':

    from argparse import ArgumentParser
    from pathlib import Path
    from matplotlib.animation import FuncAnimation
    import matplotlib.pyplot as plt
    from skfem.visuals.matplotlib import plot

    parser = ArgumentParser(description='heat equation in a rectangle')
    parser.add_argument('-g', '--gif', action='store_true', 
                        help='write animated GIF', )
    args = parser.parse_args()

    ax = plot(mesh, T_init[basis['cell'].nodal_dofs.flatten()], shading='gouraud')
    title = ax.set_title('t = 0.00')
    field = ax.get_children()[0]  # vertex-based temperature-colour
    fig = ax.get_figure()
    fig.colorbar(field)


    def update(event):
        t, u = event
        title.set_text(f'$t$ = {t:.2f}')
        field.set_array(u[basis['cell'].nodal_dofs.flatten()])

    animation = FuncAnimation(
        fig,
        update,
        evolve(0., T_init),
        repeat=False,
        interval=50,
    )
    if args.gif:
        animation.save(Path(__file__).with_suffix('.gif'), 'imagemagick')
    else:
        plt.show()
```

**Reply by abhishek-appana** [reply]
Hello @kinnala & @gdmcbain ,
Could I get help with this, please?

If needed, below is the derivation of the weak formulation and implementation of the approximation!

$\rho C_p \partial u/\partial t = k\Delta u + Q_{gen}$
Multiplying with the test function $v$ and integrating on the domain,
$\int_\Omega \rho C_p (\partial u/\partial t)vdx = \int_\Omega k\Delta u vdx +\int_\Omega Q_{gen}vdx$

Applying integration-by-parts on the first term on the right-hand side and using the boundary conditions, we get:
$\int_\Omega k\Delta u vdx = k \nabla u v |^\Omega - k\int_\Omega \nabla u . \nabla v dx = -\int_{4,6} q_{tabs}vds -\int_{rest}h(u-T_{inf})vds - k\int_\Omega \nabla u . \nabla v dx $

Substituting in the original equation:
$\int_\Omega \rho C_p (\partial u/\partial t)vdx + k\int_\Omega \nabla u . \nabla v dx + \int_{4,6} q_{tabs}vds + \int_{rest}h(u-T_{inf})vds = \int_\Omega Q_{gen}vdx$ (rest represents the edges 1,2,3,5,7,8)
Using $u = \Sigma_j \alpha_j(t) \psi_j(x)$ and $v = \psi_i(x) $:
$\int_\Omega \rho C_p \Sigma_j \alpha_j{'}\psi_j \psi_i dx + \int_\Omega k \Sigma_j \alpha_j \nabla\psi_j \nabla\psi_i dx + \int_{4,6} q_{tabs}\psi_i ds + \int_{rest} h\Sigma_j \alpha_j \psi_j \psi_i ds - \int_{rest}hT_{inf}\psi_i ds = \int_\Omega Q_{gen} \psi_i dx$

Defining new terms: $M = \int_\Omega \rho C_p\psi_j \psi_i dx, L = \int_\Omega k \nabla\psi_j \nabla\psi_i dx, O = \int_{rest} h\psi_j \psi_i ds $, and writing in matrix form,

$M \alpha' + L\alpha +  \int_{4,6} q_{tabs}\psi_i ds + O\alpha - \int_{rest}hT_{inf}\psi_i ds = \int_\Omega Q_{gen} \psi_i dx$
$M \alpha' = - L\alpha - O\alpha - \int_{4,6} q_{tabs}\psi_i ds + \int_{rest}hT_{inf}\psi_i ds + \int_\Omega Q_{gen} \psi_i dx$

Writing $\alpha' = (u-u_{old})/dt, \alpha = (u+u_{old})/2$,
$M (u-u_{old})/dt = -(L+O) (u+u_{old})/2 - \int_{4,6} q_{tabs}\psi_i ds + \int_{rest}hT_{inf}\psi_i ds + \int_\Omega Q_{gen} \psi_i dx$
$Mu - Mu_{old} = -(L+O) dt u/2 -(L+O)dt u_{old}/2 - dt\int_{4,6} q_{tabs}\psi_i ds + dt\int_{rest}hT_{inf}\psi_i ds + dt\int_\Omega Q_{gen} \psi_i dx$
$(M + (L+O)dt/2)u = (M - (L+O)dt/2)u_{old} - dt\int_{4,6} q_{tabs}\psi_i ds + dt\int_{rest}hT_{inf}\psi_i ds + dt\int_\Omega Q_{gen} \psi_i dx$

$A = M + (L+O)dt/2, B = M - (L+O)dt/2$

**Reply by gdmcbain** [reply]
I had a quick look last week but nothing jumped out at me so I can only give some general advice based on personal experience:

- The example is a bit complicated. Could you boil it down to an ["MWE"](https://en.m.wikipedia.org/wiki/Minimal_reproducible_example)?
- How I usually (always) proceed in such cases is by "baby steps": start with something that works and then add one feature at a time, getting that to work before moving on.

Sorry I can't be more immediately helpful in this case. I'm pretty sure everything required for this problem is available in skfem and that it's just going to be a matter of proceeding more gradually and carefully to the good solution.

**Reply by kinnala** [reply]
Yes, start from some simple implicit Euler solver and then try to apply those backsolve optimizations and more general methods after you are sure it works:
```python
from skfem import *
from skfem.helpers import dot, grad
import numpy as np

length = 0.5
height = 0.5
ncells = 50

T_inf = 298
Q_gen = 1000
q_tabs = 500
rho = 25
C_p = 90
k = 25
h = 10

dt = 1

mesh = (MeshQuad
        .init_tensor(np.linspace(0, length, ncells),
                     np.linspace(0, height, ncells))
        .with_boundaries({'1': lambda x: np.isclose(x[1],0),
                          '2': lambda x: np.isclose(x[0],0),
                          '3': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0) & np.less_equal(x[0],0.1),
                          '4': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.1) & np.less_equal(x[0],0.2),
                          '5': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.2) & np.less_equal(x[0],0.3),
                          '6': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.3) & np.less_equal(x[0],0.4),
                          '7': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.4) & np.less_equal(x[0],0.5),
                          '8': lambda x: np.isclose(x[0],0.5)}))

e = ElementQuad1()  # or ElementQuad1
basis = Basis(mesh, e)
fbasis_N = basis.boundary({'4', '6'})
fbasis_R = basis.boundary({'1', '2', '3', '5', '7', '8'})
fbasis = basis.boundary()


@BilinearForm
def bilinf(u, v, w):
    return rho * C_p / dt * u * v + k * dot(grad(u), grad(v))

@BilinearForm
def fbilinf(u, v, w):
    return h * u * v

@LinearForm
def linf(v, w):
    return rho * C_p / dt * w['prev'] * v + Q_gen * v

@LinearForm
def flinf_R(v, w):
    return k * T_inf * v

@LinearForm
def flinf_N(v, w):
    return - q_tabs * v


A = bilinf.assemble(basis) + fbilinf.assemble(fbasis_R)
b = flinf_R.assemble(fbasis_R) + flinf_N.assemble(fbasis_N)

t = 0
x = basis.zeros() + 298
while t < 100:
    x = solve(A, b + linf.assemble(basis, prev=x))
    t += dt
    print("t = {}".format(t))

basis.plot(x, colorbar=True).show()
```

**Reply by kinnala** [reply]
Note: I'm not much of a finite difference fellow so the above might be totally incorrect.

**Reply by abhishek-appana** [reply]
Thank you for responding @gdmcbain & @kinnala. And, thank you for your suggestions

The actual problem I am trying to solve is slightly more complicated than this. ($Q_{gen}$ term is not uniform across the geometry and the geometry is 3 dimensional). I am not sure what else I can do to simplify the example.

In terms of adding the features one by one, this is what I did before:
- I did not separate the top into 5 sections and left it to have the same boundary condition throughout.
- I checked the solution with neumann boundary conditions on all 4 sides. The solution matched with other FEM solvers
- I checked the solution with neumann boundary condition on left, bottom and right edges and the robin boundary condition on top edge. The solution matched for this case as well.

I am running into issues when I implement different boundary conditions on different sections of the top edge

@kinnala: Thank you for your suggestion. The results from this are not at all close to what I expected. I will look into using this method further and see how it goes!

**Reply by gdmcbain** [reply]
Yes, switching from Crank-Nicolson (theta=0.5) to backward Euler (theta=1) is a good idea; much more stable, much less sensitive to initial conditions. No need to rewrite the code (from e.g. ex19) just change the value of `theta`.

**Reply by kinnala** [reply]
I'd probably try reproducing the static case $$\partial_t T = 0$$ which should correspond to $$t \rightarrow \infty$$ because you have no time-dependent loading or BC, using the time-dependent solver by running it long enough, just to see if the time discretization is correct.

**Reply by gdmcbain** [reply]
> I'd probably try reproducing the static case $$\partial_t T = 0$$ which should correspond to $$t \rightarrow \infty$$ because you have no time-dependent loading or BC, using the time-dependent solver by running it long enough, just to see if the time discretization is correct.

Yes, with theta=1 ("implicit" or "backward" Euler), the time step `dt` may been taken as long as one likes, even to `numpy.Inf` if one divides through by `dt`, in which case the one backward Euler step reduces to the steady state solve.

**Reply by gdmcbain** [reply]
> The actual problem I am trying to solve is slightly more complicated than this. (&hellip;term is not uniform across the geometry and the geometry is 3 dimensional). I am not sure what else I can do to simplify the example.

Another idea here is to avoid defining the `.mesh_boundaries` via `numpy.isclose`.  This tends to be complicated and a source of errors.  A tidier way is to define the geometry and meshes in a specialized tool like [Gmsh](<https://gmsh.info/>), including [`Physical Curve`](https://gmsh.info/doc/texinfo/gmsh.html#index-Physical-curves) entities for the different parts of the mixed boundary condition, export in Gmsh's native MSH file format, and load with `Mesh.load` as in ex28:

https://github.com/kinnala/scikit-fem/blob/2bf93b7505d25978b8f9a04435bd795678a5462b/docs/examples/ex28.py#L72

Defining the geometry and mesh in something like Gmsh allows the _geometry_ to be defined exactly, including the patches, and then meshed conformally; this avoids having to try to define the boundary patches quasi-geometrically on a discrete mesh without being able to refer to where the nodes and elements of the mesh actually area.

This will become even more important in three dimensions.

**Reply by abhishek-appana** [reply]
Thank you for your suggestions @kinnala @gdmcbain.

I am exploring all these and some other ideas I have. Will update the forum!

**Comment by kinnala** [comment]
> @kinnala: Thank you for your suggestion. The results from this are not at all close to what I expected. I will look into using this method further and see how it goes!

I have never carefully studied time-dependent PDE problems other than learning the basics so I cannot help too much in the implementation or verification. Above I've attempted to use the implicit Euler method which I learned as a student long time ago but haven't used since. It is very simple but I did not verify the result in any way. There must be some test problems available for the heat equation where you can verify that you have implemented the solver correctly, unfortunately I have no time to dig through the references right now.

I must say that the package scikit-fem is primarily for the spatial discretization, which I am more familiar with, even though others have demonstrated the application of scikit-fem to the spatial discretization of a time-dependent PDE problem. I'm quite sure many other integrated solvers exist for time-dependent problems that have thoroughly tested physics modules for the heat equation.

**Reply by gdmcbain** [reply]
Yes examples 19 & 39 demonstrate the generalized trapezoidal rule for the heat equation in one and two dimensions against the exact solutions, as derived using the classical techniques of Carslaw & Jaeger's _Conduction of Heat in Solids_; no problem there.
